File: src/Scripting/Lua/decorator.lua
local function plain()
  return "alert"
end

local function audit(inner)
  return function()
    return "audit(" .. inner() .. ")"
  end
end

local function encrypt(inner)
  return function()
    return "enc(" .. inner() .. ")"
  end
end

print("base=" .. plain())
print("audit=" .. audit(plain)())
print("encrypted=" .. encrypt(plain)())
print("stacked=" .. audit(encrypt(plain))())

File: src/Scripting/Perl/decorator.pl
use strict;
use warnings;

sub plain {
    return 'alert';
}

sub audit {
    my ($inner) = @_;
    return sub { return 'audit(' . $inner->() . ')'; };
}

sub encrypt {
    my ($inner) = @_;
    return sub { return 'enc(' . $inner->() . ')'; };
}

my $component = \&plain;
my $audited = audit($component);
my $encrypted = encrypt($component);
my $stacked = audit(encrypt($component));

print 'base=' . $component->() . "\n";
print 'audit=' . $audited->() . "\n";
print 'encrypted=' . $encrypted->() . "\n";
print 'stacked=' . $stacked->() . "\n";
